src/veil/ringct/anonwallet.cpp

```diff
@@ -387,12 +387,106 @@ bool AnonWallet::GetAddressMeta(const CStealthAddress& address, CKeyID& idAccoun
     if (mapKeyPaths.count(sxAddr.GetID())) {
         auto accountpair = mapKeyPaths.at(sxAddr.GetID());
         idAccount = accountpair.first;
+        std::string strPathToKey = BIP32PathToString(accountpair.second);
+
+        //The path of the account is what provides context
+        accountpair = mapKeyPaths.at(idAccount);
         strPath = BIP32PathToString(accountpair.second);
+        strPath += strPathToKey;
     }
 
     return true;
 }
 
+CStealthAddress GenerateStealthAddressFromIndex(const CExtKey& keyStealthAddress, int nIndex)
+{
+    //Generate the extkey for spend and scan
+    BIP32Path vPathNew;
+    vPathNew.emplace_back(nIndex, true);
+    CExtKey extKeyScan = DeriveKeyFromPath(keyStealthAddress, vPathNew);
+    vPathNew.clear();
+    vPathNew.emplace_back(nIndex+1, true);
+    CExtKey extKeySpend = DeriveKeyFromPath(keyStealthAddress, vPathNew);
+    CKey keyScan = extKeyScan.key;
+    CKey keySpend = extKeySpend.key;
+    CPubKey pkSpend = keySpend.GetPubKey();
+
+    // Make a stealth address
+    CStealthAddress stealthAddress;
+    stealthAddress.SetNull();
+    stealthAddress.scan_pubkey = keyScan.GetPubKey().Raw();
+    stealthAddress.spend_pubkey = pkSpend.Raw();
+    stealthAddress.scan_secret.Set(keyScan.begin(), true);
+    stealthAddress.spend_secret_id = pkSpend.GetID();
+
+    return stealthAddress;
+}
+
+//! Return the last index in this account that has received coins
+int AnonWallet::GetLastUsedAddressIndex(const CKeyID& idAccount) const
+{
+    if (!mapAccountCounter.count(idAccount))
+        return -1;
+
+    int nLastGenerated = mapAccountCounter.at(idAccount);
+    int n = nLastGenerated - 1;
+    while (n > 0) {
+        //Regenerate address for this path
+        CExtKey keyStealthAccount;
+        if (!RegenerateKeyFromIndex(idAccount, n, keyStealthAccount))
+            return -1;
+
+        //Regenerate the stealth address and then get full address from 
+        CStealthAddress stealthAddress = GenerateStealthAddressFromIndex(keyStealthAccount, 1);
+        if (!mapStealthAddresses.count(stealthAddress.GetID()))
+            return -2;
+
+        stealthAddress = mapStealthAddresses.at(stealthAddress.GetID());
+
+        //If this address has any stealth destinations, then it is used and is the last used address.
+        if (!stealthAddress.setStealthDestinations.empty())
+            return n;
+
+        n--;
+    }
+    return -1;
+}
+
+void AnonWallet::ForgetUnusedStealthAddresses(int nBuffer)
+{
+    int nIndexLastUsed = GetLastUsedAddressIndex(idStealthAccount);
+    if (nIndexLastUsed < 0)
+        return;
+
+    // Get the last used address (this must be done after rescanning the chain)
+    int nLastGenerated = mapAccountCounter.at(idStealthAccount);
```

Comment:
Should check that mapAccountCounter contains the idStealthAccount first? Just as a precaution
`if (!mapAccountCounter.count(idStealthAccount) { return; }`

Reply:
`GetLastUsedAddressIndex()` returns negative if `idStealthAccount` is not in `mapAccountCounter`